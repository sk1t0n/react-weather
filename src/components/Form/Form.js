import React from 'react';
import FormBootstrap from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import styles from './Form.module.css';

function Form(props) {
  return (
    <FormBootstrap className={styles.Form}>
      <FormBootstrap.Group className="mb-3" controlId="city">
        <FormBootstrap.Label>City</FormBootstrap.Label>
        <FormBootstrap.Control type="text" placeholder="Enter city" />
      </FormBootstrap.Group>

      <Button variant="primary" type="submit">
        Submit
      </Button>
    </FormBootstrap>
  );
}

export { Form };
