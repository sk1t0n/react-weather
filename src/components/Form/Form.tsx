import React, { useState, useRef } from 'react';
import FormBootstrap from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import { Result } from '../Result';

type ErrorMessages = {
  [key: string]: string;
};

export type Temperature = {
  temp_c: number;
  temp_f: number;
};

export const Form: React.FC = () => {
  const API_KEY = process.env['REACT_APP_WEATHER_KEY'];
  const url = `https://api.weatherapi.com/v1/current.json?key=${API_KEY}&q=`;
  const errorMessages: ErrorMessages = {
    '400': 'City not found, maybe you entered the wrong city name. Correct the error and try again.'
  };

  let [isLoading, setLoading] = useState(false);
  let [temperature, setTemperature] = useState<Temperature | null>(null);
  let [error, setError] = useState<string | null>(null);
  let inputRef = useRef<HTMLInputElement>(null);

  async function handleSubmit(event: React.MouseEvent) {
    event.preventDefault();
    setLoading(true);
    setError(null);
    setTemperature(null);

    let value = inputRef?.current?.value;
    if (value && value.length > 0) {
      try {
        const response = await fetch(url + value);

        if (!response.ok) {
          setError(errorMessages[response.status.toString()]);
          setLoading(false);
        } else {
          const json = await response.json();
          const temp_c = json.current?.temp_c;
          const temp_f = json.current?.temp_f;
          setTemperature({ temp_c, temp_f });
          setLoading(false);
        }
      } catch(e: any) {
        setError(e.message);
        setLoading(false);
      }
    } else {
      setLoading(false);
    }
  }

  function handleReset() {
    setError(null);
    setTemperature(null);
  }

  return (
    <div className="max-width-600_center">
      <FormBootstrap className="border rounded p-4">
        <FormBootstrap.Group className="mb-3" controlId="city">
          <FormBootstrap.Label>City</FormBootstrap.Label>
          <FormBootstrap.Control ref={inputRef} type="text" placeholder="Enter city" />
        </FormBootstrap.Group>

        <Button onClick={handleSubmit} variant="primary" type="submit">
          Submit
        </Button>
        <Button onClick={handleReset} variant="warning" className="mx-3" type="reset">
          Reset
        </Button>

      </FormBootstrap>
      <Result isLoading={isLoading} error={error} temperature={temperature} />
    </div>
  );
}
