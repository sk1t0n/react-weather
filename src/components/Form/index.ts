export { Form } from './Form';
export type { Temperature } from './Form';
