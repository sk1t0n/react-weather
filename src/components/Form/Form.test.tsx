import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Form } from './Form';
import fetch from 'jest-fetch-mock';

describe('Form component', () => {
  it('check the snapshot', () => {
    const { asFragment } = render(<Form />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Testing DOM', () => {
    beforeEach(() => {
      fetch.resetMocks();
    });

    it('submit valid data', async () => {
      fetch.mockResponse(
        JSON.stringify({ temp_c: 0, temp_f: 32 }),
        { status: 200 }
      );
      fetch('http://example-api.local');

      render(<Form />);
      userEvent.type(
        screen.getByPlaceholderText('Enter city'),
        'Moscow'
      );
      userEvent.click(screen.getByText('Submit'));
      expect(await screen.findByRole('alert')).toHaveTextContent('Temperature');
    });

    it('submit invalid data', async () => {
      fetch.mockResponse('', { status: 400 });
      fetch('http://example-api.local');

      render(<Form />);
      userEvent.type(
        screen.getByPlaceholderText('Enter city'),
        'Invalid city'
      );
      userEvent.click(screen.getByText('Submit'));
      expect(await screen.findByRole('alert')).toHaveTextContent('City not found');
    });

    it('click the reset button', async () => {
      render(<Form />);
      userEvent.type(
        screen.getByPlaceholderText('Enter city'),
        'Moscow'
      );

      expect(screen.getByDisplayValue('Moscow')).toBeInTheDocument();
      userEvent.click(screen.getByText('Reset'));
      expect(screen.queryByDisplayValue('Moscow')).toBeNull();
    });
  });
});
