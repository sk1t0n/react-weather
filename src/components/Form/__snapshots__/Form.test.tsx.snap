// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component check the snapshot 1`] = `
<DocumentFragment>
  <div
    class="max-width-600_center"
  >
    <form
      class="border rounded p-4"
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="city"
        >
          City
        </label>
        <input
          class="form-control"
          id="city"
          placeholder="Enter city"
          type="text"
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Submit
      </button>
      <button
        class="mx-3 btn btn-warning"
        type="reset"
      >
        Reset
      </button>
    </form>
    <div
      class="mt-5"
    />
  </div>
</DocumentFragment>
`;
